### PR TITLE
Samsung Dex: Handle WinKey and Alt=Tab

### PR DESCRIPTION
**Pull request by alexmarian99**
Samsung dex capture of Win Key and Alt + Tab to onKeyDown and onKeyUp events for better user experience.

### CODE REVIEW

File: app/src/main/java/com/limelight/utils/SamsungDexUtils.kt [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    fun dexMetaKeyCapture(activity: Activity) {
+        runCatching {
+            val semWindowManager = Class.forName("com.samsung.android.view.SemWindowManager")
+            val getInstanceMethod = semWindowManager.getMethod("getInstance")
+            val manager = getInstanceMethod.invoke(null)
+
+            val requestMetaKeyEvent = semWindowManager.getDeclaredMethod(
+                    "requestMetaKeyEvent", ComponentName::class.java, Boolean::class.java)
+            requestMetaKeyEvent.invoke(manager, activity.componentName, true)
```
Comment (cgutman) [inline app/src/main/java/com/limelight/utils/SamsungDexUtils.kt:18]: Is this technique documented anywhere or did you figure it out yourself?

File: app/src/main/jni/moonlight-core/simplejni.c [hunk 2]
```diff
@@ -20,7 +20,7 @@ Java_com_limelight_nvstream_jni_MoonBridge_sendMousePosition(JNIEnv *env, jclass
 JNIEXPORT void JNICALL
 Java_com_limelight_nvstream_jni_MoonBridge_sendMouseMoveAsMousePosition(JNIEnv *env, jclass clazz,
         jshort deltaX, jshort deltaY, jshort referenceWidth, jshort referenceHeight) {
-    LiSendMouseMoveAsMousePositionEvent(deltaX, deltaY, referenceWidth, referenceHeight);
+    LiSendMousePositionEvent(deltaX, deltaY, referenceWidth, referenceHeight);
```
Comment (cgutman) [inline app/src/main/jni/moonlight-core/simplejni.c:23]: It looks like you somehow partially reverted b3700b5a19a8f2a6df1d4a56b1e7e053054545a3. Please undo that part.

File: app/build.gradle [hunk 3]
```diff
@@ -124,4 +125,8 @@ dependencies {
     implementation 'com.squareup.okhttp3:okhttp:3.12.13'
     implementation 'com.squareup.okio:okio:1.17.5'
     implementation 'org.jmdns:jmdns:3.5.7'
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
```
Comment (cgutman) [inline app/build.gradle:128]: Let's not pull in all this Kotlin stuff for one tiny function. Please move `dexMetaKeyCapture()` to Game.java or another Java class.

**Review by cgutman — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/limelight/utils/SamsungDexUtils.kt (hunk 1 above)
Comment (alexmarian99) [reply app/src/main/java/com/limelight/utils/SamsungDexUtils.kt:18]: Samsung didn't document that part on their developer website.
I saw Microsoft RDP handle keys on Samsung Dex, do debugging while running Microsoft RDP and figure out those changes.

Comment (alexmarian99) [reply app/src/main/java/com/limelight/utils/SamsungDexUtils.kt]: Also , it seems to capture WinKey and Alt + Tab also in Tablet mode, not only in Dex.
Would that changes make it work on other devices, not only Samsung?
I don't have other devices to test it out.

**Comment by cgutman** [comment]
Did you close this on purpose?

File: app/src/main/java/com/limelight/utils/SamsungDexUtils.kt (hunk 1 above)
Comment (cgutman) [reply app/src/main/java/com/limelight/utils/SamsungDexUtils.kt]: > Also , it seems to capture WinKey and Alt + Tab also in Tablet mode, not only in Dex.

Nice! That will be useful for Samsung devices.

> Would that changes make it work on other devices, not only Samsung?

No, `requestMetaKeyEvent()` is not part of AOSP, so probably only Samsung devices support it.

**Comment by alexmarian99** [comment]
yes, i'm trying to find an universal solution, not only for samsung , android 11 + . Got a tablet with vanilla android, and i would like firstly see if exist a better method

**Comment by cgutman** [comment]
@alexmarian99 I added your technique in 84fcd3a. Even though it's only for Samsung devices, it's better than nothing.

Can you double check that my implementation works? My old Galaxy Note 9 doesn't have the `requestMetaKeyEvent()` API.

**Comment by alexmarian99** [comment]
I'm gonna check with my devices for any bugs and come back.
I've got a better method, will come on a new pull request.
Also, i got touchpad gestures from samsung keyboard cover for tab s7 and tab s8 work. Wanna include them also in the new repo?

**Comment by cgutman** [comment]
> Also, i got touchpad gestures from samsung keyboard cover for tab s7 and tab s8 work. Wanna include them also in the new repo?

Please do them in separate pull requests.

**Comment by alexmarian99** [comment]
> 

@cgutman , can you try this [release of moonlight](https://github.com/alexmarian99/moonlight-samsung/releases/tag/v10.2) if works on your galaxy note 9 ?

**Pull request closed without merging**